tests/qnn/test_keras.py

```diff
@@ -0,0 +1,425 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Tests for the pennylane.qnn.keras module.
+"""
+import os
+import tempfile
+
+import numpy as np
+import pytest
+
+import pennylane as qml
+from pennylane.qnn.keras import KerasLayer
+
+tf = pytest.importorskip("tensorflow", minversion="2")
+
+
+@pytest.fixture
+def get_circuit(n_qubits, output_dim, interface):
+    """Fixture for getting a sample quantum circuit with a controllable qubit number and output
+    dimension. Returns both the circuit and the shape of the weights."""
+
+    dev = qml.device("default.qubit", wires=n_qubits)
+    weight_shapes = {
+        "w1": (3, n_qubits, 3),
+        "w2": (1,),
+        "w3": 1,
+        "w4": [3],
+        "w5": (2, n_qubits, 3),
+        "w6": 3,
+        "w7": 0,
+    }
+
+    @qml.qnode(dev, interface=interface)
+    def circuit(inputs, w1, w2, w3, w4, w5, w6, w7):
+        """A circuit that embeds data using the AngleEmbedding and then performs a variety of
+        operations. The output is a PauliZ measurement on the first output_dim qubits. One set of
+        parameters, w5, are specified as non-trainable."""
+        qml.templates.AngleEmbedding(inputs, wires=list(range(n_qubits)))
+        qml.templates.StronglyEntanglingLayers(w1, wires=list(range(n_qubits)))
+        qml.RX(w2[0], wires=0 % n_qubits)
+        qml.RX(w3, wires=1 % n_qubits)
+        qml.Rot(*w4, wires=2 % n_qubits)
+        qml.templates.StronglyEntanglingLayers(w5, wires=list(range(n_qubits)))
+        qml.Rot(*w6, wires=3 % n_qubits)
+        qml.RX(w7, wires=4 % n_qubits)
+        return [qml.expval(qml.PauliZ(i)) for i in range(output_dim)]
+
+    return circuit, weight_shapes
+
+
+@pytest.mark.usefixtures("get_circuit")
+@pytest.fixture
+def model(get_circuit, n_qubits, output_dim):
+    """Fixture for creating a hybrid Keras model. The model is composed of KerasLayers sandwiched
+    between Dense layers."""
+    c, w = get_circuit
+    layer1 = KerasLayer(c, w, output_dim)
+    layer2 = KerasLayer(c, w, output_dim)
+
+    model = tf.keras.models.Sequential(
+        [
+            tf.keras.layers.Dense(n_qubits),
+            layer1,
+            tf.keras.layers.Dense(n_qubits),
+            layer2,
+            tf.keras.layers.Dense(output_dim),
+        ]
+    )
+
+    return model
+
+
+def indicies_up_to(n_max):
+    """Returns an iterator over the number of qubits and output dimension, up to value n_max.
+    The output dimension never exceeds the number of qubits."""
+    a, b = np.tril_indices(n_max)
+    return zip(*[a + 1, b + 1])
+
+
+@pytest.mark.usefixtures("get_circuit")
+class TestKerasLayer:
+    """Unit tests for the pennylane.qnn.keras.KerasLayer class."""
+
+    @pytest.mark.parametrize("interface", ["tf"])
```

Comment:
Is this parametrization needed?

Reply:
While this wasn't a 'proper' usage of PyTest parametrized fixtures, more of a side-effect, as of the latest bugfix version I think they now count this as expected behaviour (as so many projects opened bug reports when they removed it, see https://github.com/pytest-dev/pytest/issues/6909).

So happy to keep this as-is @trbromley! Maybe just add a comment?
```suggestion
    # the get_circuit fixture requires that interface be defined
    @pytest.mark.parametrize("interface", ["tf"])
```

Reply:
Thanks @josh146, added a comment.